### PR TITLE
[ExpandLargeFpConvert] Scalarize vector types.

### PR DESCRIPTION
**Pull request by bevin-hansson**
expand-large-fp-convert cannot handle vector types.
If overly large vector element types survive into
isel, they will likely be scalarized there, but since
isel cannot handle scalar integer types of that size,
it will assert.

Handle vector types in expand-large-fp-convert by
scalarizing them and then expanding the scalar type
operation. For large vectors, this results in a
*massive* code expansion, but it's better than
asserting.

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
What's the difficulty supporting vectors in the expansion directly? I assume it requires simple bookkeeping adjustments to handle them?

**Comment by bevin-hansson** [comment]
> What's the difficulty supporting vectors in the expansion directly? I assume it requires simple bookkeeping adjustments to handle them?

It's the same issue as in #86959. The presence of control flow makes it a bit hard to emit vector code that matches the original.

File: llvm/lib/CodeGen/ExpandLargeFpConvert.cpp [hunk 1]
```diff
@@ -568,8 +568,30 @@ static void expandIToFP(Instruction *IToFP) {
   IToFP->eraseFromParent();
 }
 
+static void scalarize(Instruction *I, SmallVectorImpl<Instruction *> &Replace) {
+  VectorType *VTy = cast<VectorType>(I->getType());
+  assert(!VTy->isScalableTy() && "Tried to scalarize scalable vector!");
```
Comment (arsenm) [inline llvm/lib/CodeGen/ExpandLargeFpConvert.cpp:573]: Just cast<FixedVectorType>

**Review by arsenm — COMMENTED** [review]
lgtm with nit

**Comment by bevin-hansson** [comment]
I'll push this tomorrow.

**Review by arsenm — APPROVED** [review]

**Comment by arsenm** [comment]
I do think we need a common scalarization utility. Also, I don't really see the point of having the expand-large-div and expand-large-fp pass being separate. They have identical purpose

**Pull request merged**